File: java/lib/template.ts
import {
  type Artifact,
  type ValDefs,
  type ConditionalVal,
  sourceUrl,
  extractDump,
} from '@torba/core';
import type { OsName, OsArch, Ruleset } from '@torba/rules';
import {
  resolveOpenjdk,
  type JavaPlatform,
  type JavaRelease,
  type ResolveOpenjdkOptions,
} from './resolver';

export interface JavaOptions {
  /**
   * OpenJDK version. Accepts:
   *   - Major:        `'21'` — resolves to the latest GA for that major.
   *   - Full version: `'21.0.11+10'` — exact Adoptium release name (`jdk-…` prefix and `-LTS` suffix are tolerated).
   */
  version: string;
  /** Distribution to fetch from. Only `'openjdk'` (Eclipse Temurin) is supported today. */
  vendor?: 'openjdk';
  /** Override the platform set. Default covers linux/osx/windows × x86_64+aarch64. */
  platforms?: readonly JavaPlatform[];
  /** Optional override for the Adoptium API base URL. */
  apiBase?: string;
}

export interface JavaTemplate {
  /** Per-platform JDK archives, scoped by OS+arch rules and extracted on first install. */
  artifacts: Artifact[];
  /** `java_home` and `java_bin` per OS — spread into your loader's vars. */
  vars: ValDefs;
  /** Resolved release metadata. */
  release: JavaRelease;
}

function osArchRuleset(os: OsName, arch: OsArch): Ruleset {
  return [
    { action: 'allow', os: { name: os } },
    { action: 'allow', os: { arch } },
  ];
}

function osRuleset(os: OsName): Ruleset {
  return [{ action: 'allow', os: { name: os } }];
}

/**
 * Build a torba template fragment that auto-installs an OpenJDK runtime
 * (Eclipse Temurin) and exposes `${java_home}` + `${java_bin}` vars.
 *
 * Each platform's archive is emitted as its own `Artifact` with an
 * OS+arch rule, so only the matching binary downloads at install time.
 * The archive is extracted into `${root}/runtimes/jdk-<major>/`, with
 * the Adoptium release directory as the immediate child (e.g.
 * `jdk-21.0.11+10/`). On macOS, `${java_home}` includes the
 * `/Contents/Home` suffix that Mac JDK bundles use.
 *
 * Spread the result into your loader's vars + artifacts:
 *
 * ```ts
 * const jav = await resolveJava({ version: '21' });
 * return {
 *   artifacts: [lw.artifacts, jav.artifacts],
 *   vars: { ...lw.vars, ...jav.vars },
 *   command: lw.command, // command.command is `${java_bin}` already
 * };
 * ```
 */
export async function resolveJava(options: JavaOptions): Promise<JavaTemplate> {
  if (options.vendor && options.vendor !== 'openjdk') {
    throw new Error(
      `@torba/java: vendor '${options.vendor}' is not yet supported (only 'openjdk').`,
    );
  }
  const release = await resolveOpenjdk(options.version, {
    platforms: options.platforms,
    apiBase: options.apiBase,
  } satisfies ResolveOpenjdkOptions);

  const javaRoot = `\${root}/runtimes/jdk-${release.major}`;
  const archiveDir = `${javaRoot}/.cache`;
  const extractInto = javaRoot;

  const artifacts: Artifact[] = release.binaries.map((b) => ({
    path: `${archiveDir}/${b.filename}`,
    source: sourceUrl(b.url),
    size: b.size,
    rules: osArchRuleset(b.platform.os, b.platform.arch),
    integrity: { sha256: b.sha256 },
    extract: [extractDump(extractInto, { excludes: [] })],
  }));

  // `java_home` only varies by OS (Linux/Windows have no suffix; macOS
  // bundles add `/Contents/Home`). Both x86_64 and aarch64 archives
  // extract to the same top-level directory, so we don't need to split
  // by arch here.
  const seenOses = new Set(release.binaries.map((b) => b.platform.os));
  const javaHomeArms: ConditionalVal[] = [];
  const javaBinArms: ConditionalVal[] = [];
  for (const os of seenOses) {
    const platform = release.binaries.find(
      (b) => b.platform.os === os,
    )!.platform;
    const home = `${javaRoot}/${release.extractDir}${platform.homeSuffix}`;
    javaHomeArms.push({ value: home, rules: osRuleset(os) });
    const exe = os === 'windows' ? 'java.exe' : 'java';
    javaBinArms.push({
      value: `\${java_home}/bin/${exe}`,
      rules: osRuleset(os),
    });
  }

  const vars: ValDefs = {
    java_home: javaHomeArms,
    java_bin: javaBinArms,
  };

  return { artifacts, vars, release };
}

File: crates/opys-java/tests/template.rs
//! The vendor-agnostic half: a resolved release in, artifacts and vars out.
//! Pure, so most of this needs no server at all — only the vendor-dispatch
//! cases do.

mod common;

use common::{Reply, Request, TestServer};
use opys_core::{
    ExtractRule, HashEntry, Integrity, MojangRule, OsArch, OsConstraint, OsName, RuleAction,
    Source, ValDef,
};
use opys_java::{
    build_java, java_template, resolve_java, JavaOptions, JavaVendor, Platform, SupportedArch,
    VendorBinary, VendorRelease,
};

const LINUX_X64: Platform = Platform {
    os: OsName::Linux,
    arch: SupportedArch::X86_64,
};
const MAC_AARCH64: Platform = Platform {
    os: OsName::Osx,
    arch: SupportedArch::Aarch64,
};
const WIN_X64: Platform = Platform {
    os: OsName::Windows,
    arch: SupportedArch::X86_64,
};

fn binary(platform: Platform, filename: &str) -> VendorBinary {
    VendorBinary {
        platform,
        filename: filename.to_owned(),
        url: format!("https://example.invalid/{filename}"),
        size: 100,
        sha256: Some("cafe".to_owned()),
        discovery: None,
    }
}

fn release(binaries: Vec<VendorBinary>) -> VendorRelease {
    VendorRelease {
        label: "Temurin 21.0.11+10".to_owned(),
        major: 21,
        binaries,
    }
}

fn arms(def: &ValDef) -> &[opys_core::ConditionalVal] {
    match def {
        ValDef::Arms(arms) => arms,
        ValDef::Flat(_) => panic!("expected conditional arms"),
    }
}

// ── artifacts ─────────────────────────────────────────────────────────────

#[test]
fn emits_one_artifact_per_resolved_binary() {
    let template = java_template(release(vec![
        binary(LINUX_X64, "linux.tar.gz"),
        binary(MAC_AARCH64, "mac.tar.gz"),
    ]));
    assert_eq!(template.artifacts.len(), 2);
}

#[test]
fn places_archives_under_the_runtime_dir_as_a_url_source() {
    let template = java_template(release(vec![binary(LINUX_X64, "linux.tar.gz")]));
    let artifact = &template.artifacts[0];
    assert_eq!(artifact.path, "${java_runtime_dir}/linux.tar.gz");
    assert_eq!(
        artifact.source,
        Source::Url {
            url: "https://example.invalid/linux.tar.gz".to_owned()
        }
    );
    assert_eq!(artifact.size, Some(100));
    assert_eq!(
        artifact.integrity,
        Some(Integrity::One(HashEntry::Sha256 {
            sha256: "cafe".to_owned()
        }))
    );
}

#[test]
fn scopes_each_artifact_with_an_os_and_arch_ruleset() {
    let template = java_template(release(vec![binary(MAC_AARCH64, "mac.tar.gz")]));
    assert_eq!(
        template.artifacts[0].rules,
        vec![
            MojangRule::Os {
                action: RuleAction::Allow,
                os: OsConstraint {
                    name: Some(OsName::Osx),
                    ..Default::default()
                }
            },
            MojangRule::Os {
                action: RuleAction::Allow,
                os: OsConstraint {
                    arch: Some(OsArch::Aarch64),
                    ..Default::default()
                }
            },
        ]
    );
}

#[test]
fn extracts_with_a_glob_strip_into_the_major_versioned_runtime_dir() {
    let template = java_template(release(vec![binary(LINUX_X64, "linux.tar.gz")]));
    let extract = template.artifacts[0]
        .extract
        .as_ref()
        .expect("an extract rule");
    let ExtractRule::Scan(scan) = &extract[0] else {
        panic!("expected a scan rule, got {:?}", extract[0]);
    };
    assert_eq!(scan.matches, "*");
    assert_eq!(scan.into, "${java_runtime_dir}/jdk-21");
    // The archive's own top-level directory embeds a build id no resolver can
    // predict, so it is stripped by shape rather than by name.
    assert_eq!(scan.strip.as_deref(), Some(["*/".to_owned()].as_slice()));
}

#[test]
fn carries_a_discovery_fallback_onto_the_artifact() {
    let mut b = binary(LINUX_X64, "linux.tar.gz");
    b.sha256 = None;
    b.discovery = Some(opys_core::Discovery {
        integrity: Some(opys_core::IntegrityProbes {
            header: None,
            url: Some(opys_core::HashRef::Sha256 {
                sha256: "${url}.sha256".to_owned(),
            }),
        }),
        size: None,
    });
    let template = java_template(release(vec![b]));
    let artifact = &template.artifacts[0];
    assert!(artifact.integrity.is_none());
    assert!(artifact.discovery.is_some());
}

// ── vars ──────────────────────────────────────────────────────────────────

#[test]
fn defines_the_runtime_dir_under_root() {
    let template = java_template(release(vec![binary(LINUX_X64, "linux.tar.gz")]));
    assert_eq!(
        template.vars["java_runtime_dir"],
        ValDef::Flat("${root}/runtimes".to_owned())
    );
}

#[test]
fn builds_a_suffix_free_java_home_on_linux() {
    let template = java_template(release(vec![binary(LINUX_X64, "linux.tar.gz")]));
    let home = arms(&template.vars["java_home"]);
    assert_eq!(home.len(), 1);
    assert_eq!(home[0].value, "${java_runtime_dir}/jdk-21");
}

#[test]
fn appends_contents_home_on_macos() {
    let template = java_template(release(vec![binary(MAC_AARCH64, "mac.tar.gz")]));
    let home = arms(&template.vars["java_home"]);
    assert_eq!(home[0].value, "${java_runtime_dir}/jdk-21/Contents/Home");
}

#[test]
fn points_java_bin_at_bin_java_off_windows() {
    let template = java_template(release(vec![binary(LINUX_X64, "linux.tar.gz")]));
    let bin = arms(&template.vars["java_bin"]);
    assert_eq!(bin.len(), 1);
    assert_eq!(bin[0].value, "${java_home}/bin/java");
}

#[test]
fn gates_the_windows_console_binary_behind_a_feature() {
    let template = java_template(release(vec![binary(WIN_X64, "win.zip")]));
    let bin = arms(&template.vars["java_bin"]);
    assert_eq!(bin.len(), 2);

    // Default arm: no console window.
    assert_eq!(bin[0].value, "${java_home}/bin/javaw.exe");
    assert_eq!(bin[0].rules[1].action(), RuleAction::Disallow);
    // Opt-in arm: a console to watch stdout/stderr.
    assert_eq!(bin[1].value, "${java_home}/bin/java.exe");
    assert_eq!(bin[1].rules[1].action(), RuleAction::Allow);

    // Exactly one arm is active for any feature state.
    let windows = opys_core::OsOptions {
        name: "windows".to_owned(),
        version: "10".to_owned(),
        arch: "x86_64".to_owned(),
    };
    for feats in [vec![], vec!["java_console".to_owned()]] {
        let active = bin
            .iter()
            .filter(|arm| opys_core::satisfies_ruleset(&arm.rules, &windows, &feats).unwrap())
            .count();
        assert_eq!(active, 1, "feats: {feats:?}");
    }
}

#[test]
fn emits_one_arm_per_distinct_os_not_per_arch() {
    let template = java_template(release(vec![
        binary(LINUX_X64, "linux-x64.tar.gz"),
        binary(
            Platform {
                os: OsName::Linux,
                arch: SupportedArch::Aarch64,
            },
            "linux-arm.tar.gz",
        ),
        binary(MAC_AARCH64, "mac.tar.gz"),
    ]));
    assert_eq!(template.artifacts.len(), 3);
    assert_eq!(arms(&template.vars["java_home"]).len(), 2);
    assert_eq!(arms(&template.vars["java_bin"]).len(), 2);
}

#[test]
fn passes_the_release_metadata_through() {
    let resolved = release(vec![binary(LINUX_X64, "linux.tar.gz")]);
    let template = java_template(resolved.clone());
    assert_eq!(template.release, resolved);
}

// ── vendor dispatch ───────────────────────────────────────────────────────

fn temurin_body() -> String {
    r#"[{"release_name":"jdk-21.0.11+10","version_data":{"major":21},"binaries":[
        {"architecture":"x64","os":"linux","image_type":"jdk","jvm_impl":"hotspot",
         "package":{"checksum":"abc","link":"https://x/a.tar.gz","name":"a.tar.gz","size":1}}]}]"#
        .to_owned()
}

fn zulu_body() -> String {
    r#"[{"name":"z.tar.gz","download_url":"https://x/z.tar.gz","size":2,
         "sha256_hash":"def","java_version":[21,0,5],"distro_version":[21,38,21]}]"#
        .to_owned()
}

fn graalvm_body() -> String {
    r#"[{"tag_name":"jdk-21.0.2","prerelease":false,"draft":false,"assets":[
        {"name":"g_linux-x64_bin.tar.gz","size":3,
         "browser_download_url":"https://x/g.tar.gz","digest":"sha256:ghi"}]}]"#
        .to_owned()
}

/// One server that answers whichever vendor's endpoint is asked for, so the
/// test asserts on the dispatch rather than on the fixture.
fn any_vendor() -> impl Fn(&Request) -> Reply {
    |req: &Request| {
        if req.target.starts_with("/assets/") {
            Reply::json(temurin_body())
        } else if req.target.starts_with("/zulu/packages/") {
            Reply::json(zulu_body())
        } else if req.target.starts_with("/repos/") {
            Reply::json(graalvm_body())
        } else {
            Reply::status(404)
        }
    }
}

fn options(server: &TestServer, vendor: Option<JavaVendor>) -> JavaOptions {
    JavaOptions {
        version: "21".to_owned(),
        vendor,
        platforms: Some(vec![LINUX_X64]),
        api_base: Some(server.base.clone()),
        token: None,
    }
}

#[test]
fn defaults_to_temurin_when_no_vendor_is_named() {
    let server = TestServer::start(any_vendor());
    let template = resolve_java(&options(&server, None)).unwrap();
    assert_eq!(template.release.label, "Temurin 21.0.11+10");
    assert!(server.targets()[0].starts_with("/assets/"));
}

#[test]
fn dispatches_zulu_to_the_azul_api() {
    let server = TestServer::start(any_vendor());
    let template = resolve_java(&options(&server, Some(JavaVendor::Zulu))).unwrap();
    assert_eq!(template.release.label, "Zulu 21.38.21 (JDK 21.0.5)");
    assert!(server.targets()[0].starts_with("/zulu/packages/"));
}

#[test]
fn dispatches_graalvm_to_the_github_releases_api() {
    let server = TestServer::start(any_vendor());
    let template = resolve_java(&options(&server, Some(JavaVendor::Graalvm))).unwrap();
    assert_eq!(template.release.label, "GraalVM CE 21.0.2");
    assert!(server.targets()[0].starts_with("/repos/"));
}

// ── the plugin's contribution ─────────────────────────────────────────────

#[test]
fn builds_a_contribution_owning_the_java_vars_and_the_bin_launch_group() {
    let server = TestServer::start(any_vendor());
    let build = build_java(&options(&server, None)).unwrap();

    assert_eq!(build.output.name, "java");
    assert_eq!(build.release.label, "Temurin 21.0.11+10");

    let contribution = &build.output.contribution;
    assert_eq!(contribution.artifacts.len(), 1);
    assert!(contribution.vars.contains_key("java_home"));
    assert!(contribution.vars.contains_key("java_bin"));
    assert_eq!(
        contribution.launch["bin"],
        opys_dev::LaunchFragment::Text("${java_bin}".to_owned())
    );
    assert_eq!(
        contribution.envs["JAVA_HOME"],
        ValDef::Flat("${java_home}".to_owned())
    );
}
